core/src/main/java/de/unistuttgart/iaas/amyassist/amy/core/speech/data/SpeechRecognitionStateVariables.java

```diff
@@ -0,0 +1,284 @@
+
+/**
+ * Service that contains all SpeechRecognition corresponding data
+ * 
+ * @author Kai Menzel
+ */
+@Service
+public class SpeechRecognitionStateVariables {
+
+	@Reference
+	private Logger logger;
+
+	@Reference
+	private OutputImpl output;
+
+	@Reference
+	private GrammarObjectsCreator creator;
+
+	@Reference
+	private SpeechInputHandler inputHandler;
+
+	// State Data Variables
+	private ListeningState currentListeningState = ListeningState.NOT_LISTENING;
+	private Grammar currentGrammar = Grammar.NONE;
+
+	private AudioInputStream ais = createNewAudioInputStream();
+
+	// Predefined ResultHandler
+	private MainGrammarSpeechResultHandler mainResultHandler;
+	private TempGrammarSpeechResultHandler tempResultHandler;
+
+	// Predefined Recognizers
+	private Thread mainRecognizer;
+
+	// -------------------------------------------------------
+
+	/**
+	 * Setter
+	 * 
+	 * @param mainResultHandler
+	 *            handles Result of Main Recognizer
+	 */
+	public void setMainResultHandler(MainGrammarSpeechResultHandler mainResultHandler) {
+		this.mainResultHandler = mainResultHandler;
+	}
+
+	/**
+	 * Setter
+	 * 
+	 * @param tempResultHandler
+	 *            handles Result of Temp Recognizer
+	 */
+	public void setTempResultHandler(TempGrammarSpeechResultHandler tempResultHandler) {
+		this.tempResultHandler = tempResultHandler;
+	}
+
+	/**
+	 * Creates the Main Recognizer
+	 */
+	public void init() {
+		this.mainRecognizer = new Thread(new SpeechRecognizer(Grammar.MAIN, this.mainResultHandler, this.ais));
+	}
+
+	// -------------------------------------------------------
+
+	/**
+	 * Change the Listening state
+	 * 
+	 * @param state
+	 *            Multi, Single, Not_Listening
+	 */
+	public void setListeningState(ListeningState state) {
+		switch (state) {
+		case MULTI_CALL_LISTENING:
+			this.output.voiceOutput("waking up");
+			break;
+		case SINGLE_CALL_LISTENING:
+			this.output.soundOutput(Sounds.SINGLE_CALL_START_BEEP);
+			break;
+		case NOT_LISTENING:
+			switch (this.currentListeningState) {
+			case MULTI_CALL_LISTENING:
+				this.output.voiceOutput("now sleeping");
+				break;
+			case SINGLE_CALL_LISTENING:
+				this.output.soundOutput(Sounds.SINGLE_CALL_STOP_BEEP);
+				break;
+			default:
+				break;
+			}
+			break;
+		default:
+			break;
+		}
+		this.currentListeningState = state;
+	}
+
+	/**
+	 * Getter
+	 * 
+	 * @return Current ListeningState
+	 */
+	public ListeningState getListeningState() {
+		return this.currentListeningState;
+	}
+
+	/**
+	 * Change the Current Grammar
+	 * 
+	 * @param currentGrammar
+	 *            Main Temp None
+	 */
+	public void setCurrentGrammar(Grammar currentGrammar) {
+		this.currentGrammar = currentGrammar;
+		switch (currentGrammar) {
+		case MAIN:
+			this.mainRecognizer.start();
+			break;
+		case TEMP:
+			(new Thread(new SpeechRecognizer(Grammar.TEMP, this.tempResultHandler, this.ais))).start();
```

Comment:
How is ensured, that the main recognizer is running anymore?

Reply:
All Loops of the SpeechRecognizer Objects check if the current GrammarState is the State matching the Recognizer. -> The MainRecognizer While breaks if the currentGrammar is not MAIN